### PR TITLE
driver-only ECDSA: enable ECDSA-based TLS 1.2 key exchanges

### LINKED ISSUE #6862: driver-only ECDSA: enable ECDSA-based TLS 1.2 key exchanges
**Context:** https://github.com/Mbed-TLS/mbedtls/issues/6839.

This task is to finish off TL-ecdsa.TLS by no longer disabling ` MBEDTLS_KEY_EXCHANGE_ECDHE_ECDSA_ENABLED` and `MBEDTLS_KEY_EXCHANGE_ECDHE_ECDSA_ENABLED` in `config_psa_crypto_config_ecdsa_use_psa` and getting testing parity.

It might follow as a consequence of previous tasks, or require more work similar to #6860.
**Note** = the description of issue #6860 is probably useful for addressing this one

**Depends on:** #6861

### PR DESCRIPTION
**Pull request by valeriosetti**
## Description

This PR re-enables ECDSA based key exchanges that were temporary disabled during previous developments of accelerated ECDSA test parity (see #6861 and previous).

Depends on #7082 

Resolves #6862 

## Gatekeeper checklist

- [ ] **changelog** not required
- [ ] **backport** not required
- [ ] **tests** not required

### CODE REVIEW

**Comment by valeriosetti** [comment]
Opening as a draft because there are failures with `gnutls-cli` locally that I want to test with the CI

**Comment by mpg** [comment]
The CIs now both fail in "Results analysis" which is `analyze_outcomes.py`. As a reminder, by clicking the link from github, we get to a truncated view which doesn't include the actual error messages (we just have too many outcome files to collect from the previous stage), but we can view the full log and scroll to the end for errors by clicking one of the two icons on above the table on the right.

image

**Comment by valeriosetti** [comment]
> The CIs now both fail in "Results analysis" which is `analyze_outcomes.py`

Yes, it took me a while to understand what was the problem, but then I downloaded the full log from Jenkins at jumping at the end it was quite clear. I am working on that ;)

**Comment by valeriosetti** [comment]
Rebased after the last update to #7082
Attached there are the results of `outcome-analysis.sh`
[outcome-analysis-results.txt](https://github.com/Mbed-TLS/mbedtls/files/10802677/outcome-analysis-results.txt)

**Comment by valeriosetti** [comment]
Rebased to align with latest #7082 updates

File: include/mbedtls/config_psa.h
Comment (mpg) [inline include/mbedtls/config_psa.h:313]: My first reaction was "can we really define macros in the `PSA_` namespace like that? Isn't that namespace reserved for the PSA spec?" then I saw the context that already defines lots of such macros, so apparently it's OK.

So instead, I'll just ask that for consistency with the surrounding macros: `PSA_HAVE` not `PSA_HAS` :)

**Review by mpg — COMMENTED** [review]
Apart from one naming issue, looks good to me if the CI is happy.

File: include/mbedtls/config_psa.h
Comment (valeriosetti) [reply include/mbedtls/config_psa.h:313]: > then I saw the context that already defines lots of such macros

eheheheh I just thought that I needed a macro like that one for better readability of the code so I went looking for "similar" ideas in the main header files and I found those ones :)
Of course your suggestion is absolutely right!

Since the branch has to be rebased anyway (due to the dependency from #7082) if you don't mind I would like to wait for the CI to complete before applying this change. I propose this way because this morning the CI was complaining for a conflict in `ssl_tls.c` for this PR. I tried to verify it locally by trying to merge into `development`, but I didn't get the same problem. Therefore I rebased on #7082 (which I know that is not aligned to development, but it didn't suffer from this problem this morning) to see if the conflict is still there

Comment (mpg) [reply include/mbedtls/config_psa.h:313]: I just merged #7082 and the CI here came back green (after only 4h, which these days is pretty good I guess), so I think you can now rebase - and fix the typo that was left out of #7082.

Comment (gilles-peskine-arm) [reply include/mbedtls/config_psa.h]: > can we really define macros in the PSA_ namespace like that? Isn't that namespace reserved for the PSA spec?

This is a long-time-unresolved issue: we don't know what prefix to use for things that are not standard PSA definitions, but are not related to Mbed TLS legacy crypto APIs or Mbed TLS X.509 and TLS APIs.

Using a `PSA/psa` prefix for nonstandard stuff is acceptable as long as we ensure that it never conflicts with an official PSA definition. So if PSA adopts one of those names, we'll have to rename our own. And if we make those a stable interface, we rely on our ability to influence PSA specifications. Fortunately that's rarely the case — most of these names are either not part our API or designated as experimental — but a few names have crept up (in `psa/crypto_extra.h`) which are stable Mbed TLS APIs that were on PSA standardization track but PSA didn't adopt the feature.

File: library/ssl_tls.c
Comment (AndrzejKurek) [inline library/ssl_tls.c:5000]:
```suggestion
#endif /* MBEDTLS_PK_CAN_ECDSA_SOME && MBEDTLS_HAS_ALG_SHA_512_VIA_MD_OR_PSA_BASED_ON_USE_PSA &&
```

**Review by AndrzejKurek — COMMENTED** [review]
As I understand removing
```
    scripts/config.py unset MBEDTLS_KEY_EXCHANGE_ECDHE_ECDSA_ENABLED
    scripts/config.py unset MBEDTLS_KEY_EXCHANGE_ECDH_ECDSA_ENABLED
```
from `component_test_psa_crypto_config_accel_ecdsa` will be a part of a different task?

File: library/ssl_tls.c
Comment (valeriosetti) [reply library/ssl_tls.c:5000]: Noted! In case no other change is required for this PR, I will solve it in a following one (ex: the one that I will create for addressing #7148)

**Review by mpg — APPROVED** [review]
LGTM

**Comment by mpg** [comment]
@AndrzejKurek If you're happy for this to be merge as-is, please check the "approve" box. (I think that's the case, but I prefer to be 100% sure before merging.)

**Comment by AndrzejKurek** [comment]
> @AndrzejKurek If you're happy for this to be merge as-is, please check the "approve" box. (I think that's the case, but I prefer to be 100% sure before merging.)

Sure, I'm just still testing the changes locally, I haven't approved this PR yet :)

**Comment by mpg** [comment]
My mistake, glad I asked, then :)

**Review by AndrzejKurek — APPROVED** [review]

File: include/mbedtls/check_config.h
Comment (AndrzejKurek) [inline include/mbedtls/check_config.h:286]: OK, I think I know what's the issue: 
if in `component_test_psa_crypto_config_accel_ecdsa` I remove
```
    scripts/config.py unset MBEDTLS_KEY_EXCHANGE_ECDHE_ECDSA_ENABLED
    scripts/config.py unset MBEDTLS_KEY_EXCHANGE_ECDH_ECDSA_ENABLED
```
everything builds and tests fine (as the test suites using ECDHE_ECDSA are skipped, there's no `MBEDTLS_PK_CAN_ECDSA_SOME`), ssl-server2 and ssl-client2 report that they can use ECDHE-ECDSA ciphersuites, but the handshake fails.
Is it a gap in the dependencies here?

Comment (mpg) [reply include/mbedtls/check_config.h:286]: Ah, good catch! Indeed now that you mention it the ECDSA part of the dependency should be:
```c
((!defined(MBEDTLS_USE_PSA_CRYPTO) && defined(MBEDTLS_ECDSA_C)) || \
 (defined(MBEDTLS_USE_PSA_CRYPTO) && defined(PSA_HAVE_FULL_ECDSA)))
```
Wow, sorry for missing that and thanks for trying things out and catching it!

I wanted to do a final ECDSA PR to add a ChangeLog entry, I'll fix that there too.

Comment (AndrzejKurek) [reply include/mbedtls/check_config.h:286]: Yep, that's the second time that that tricky pair of `MBEDTLS_USE_PSA_CRYPTO` vs `!MBEDTLS_USE_PSA_CRYPTO` catches us :) Thanks!

Comment (valeriosetti) [reply include/mbedtls/check_config.h:286]: Ok, it took me a while to understand the point here, but I agree with you now :)
@AndrzejKurek thanks for the hint!

@mpg you can assign the new issue to me so I will complete this part

Comment (valeriosetti) [reply include/mbedtls/check_config.h:286]: Or can I simply create a pull request to fix this part without an associated issue?

Comment (mpg) [reply include/mbedtls/check_config.h:286]: Ah, sorry, I had already started working on it before I was your comment.

But in general yes, it's OK to create a PR without an issue - we just need to add a size label to the PR and add it to the EPICs board in the correct EPIC. (PRs linked to an issue don't need that as they show on the board attached to their issue.)

Comment (valeriosetti) [reply include/mbedtls/check_config.h:286]: > Ah, sorry, I had already started working on it before I was your comment.

No problem! Thanks for taking care of it ;)

Comment (mpg) [reply include/mbedtls/check_config.h:286]: https://github.com/Mbed-TLS/mbedtls/pull/7245

**Pull request merged**
